File: src/main/typescript/Game.ts
import { GroupId } from "./Group"

export type GameId = string

export interface NewGame {
  groupId: GroupId,
  player1: string,
  player2: string,
  score1: number,
  score2: number,
  isDeleted: boolean
}

export interface Game extends NewGame {
  id: GameId,
  playedAt: string
}

export interface Rating {
  rating: number,
  deviation: number
}

export interface RatedGame {
  game: Game,
  playerRatings: Record<string, Rating>
}

export interface RatingRange {
  min: number,
  max: number,
  radius: number,
  length: number
}

export function ratingRange(
  { rating, deviation }: Rating,
  multiplier: number = 1.5
): RatingRange {
  const radius = multiplier * deviation
  return {
    radius,
    min: rating - radius,
    max: rating + radius,
    length: 2 * radius
  }
}

File: src/main/scala/leaderboard/game/Game.scala
package leaderboard.game

import leaderboard.common.Id
import leaderboard.group.Group

import io.circe.{Decoder, Encoder}
import io.circe.generic.semiauto._

import java.time.LocalDateTime
import scala.util.{Failure, Success, Try}

final case class Game(
  id: Id[Game],
  groupId: Id[Group],
  playedAt: LocalDateTime,
  player1: String,
  player2: String,
  score1: Double,
  score2: Double,
  isDeleted: Boolean
)

object Game {
  implicit val encoder: Encoder[Game] = deriveEncoder
}

final case class NewGame(
  groupId: Id[Group],
  player1: String,
  player2: String,
  score1: Double,
  score2: Double,
  isDeleted: Boolean
) {

  def forInsert(playedAt: LocalDateTime): Game =
    Game(Id(0), groupId, playedAt, player1, player2, score1, score2, isDeleted)

  def update(game: Game): Try[Game] =
    if (groupId == game.groupId) {
      Success(Game(game.id, game.groupId, game.playedAt, player1, player2, score1, score2, isDeleted))
    } else {
      Failure(new IllegalArgumentException("The group of a game cannot be changed"))
    }
}

object NewGame {
  implicit val decoder: Decoder[NewGame] = deriveDecoder
}
